### PR TITLE
refactoring testpmd operator to operator-sdk 1.33.0

### PR DESCRIPTION
**Pull request by capolrik**
After Testpmd-lb-operator and cnf-mac-app-operator, here it comes the refactoring of testpmd-operator

### CODE REVIEW

File: testpmd-operator/config/manager/manager.yaml [hunk 1]
```diff
@@ -1,63 +1,104 @@
...
   replicas: 1
   template:
     metadata:
+      annotations:
+        kubectl.kubernetes.io/default-container: manager
       labels:
         control-plane: controller-manager
-        example-cnf-type: testpmd-operator
     spec:
+      # TODO(user): Uncomment the following code to configure the nodeAffinity expression
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:39]: I'd remove these comments

File: testpmd-operator/config/manager/manager.yaml [hunk 2]
```diff
@@ -1,63 +1,104 @@
...
+      #               - arm64
+      #               - ppc64le
+      #               - s390x
+      #           - key: kubernetes.io/os
+      #             operator: In
+      #             values:
+      #               - linux
       securityContext:
         runAsNonRoot: true
+        # TODO(user): For common cases that do not require escalating privileges
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:61]: I'd remove these comments

File: testpmd-operator/config/manager/manager.yaml [hunk 3]
```diff
@@ -1,63 +1,104 @@
...
+            port: 6789
+          initialDelaySeconds: 15
+          periodSeconds: 20
+        readinessProbe:
+          httpGet:
+            path: /readyz
+            port: 6789
+          initialDelaySeconds: 5
+          periodSeconds: 10
+        # TODO(user): Configure the resources accordingly based on the project requirements.
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:94]: These comments can be probably removed

File: testpmd-operator/config/manager/manager.yaml [hunk 4]
```diff
@@ -1,63 +1,104 @@
...
-            periodSeconds: 20
-          readinessProbe:
-            httpGet:
-              path: /readyz
-              port: 6789
-            initialDelaySeconds: 5
-            periodSeconds: 10
-          # operator-sdk is not implementing startup probe up to now, using liveness probe endpoint
-          # while this feature is developed
-          startupProbe:
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:55]: do not remove startupProbe (comment above is not needed)

File: testpmd-operator/config/manager/manager.yaml [hunk 5]
```diff
@@ -1,63 +1,104 @@
...
-          args:
-            - "--enable-leader-election"
-            - "--leader-election-id=testpmd-operator"
-            - "--zap-encoder=console"
-          image: controller:latest
-          imagePullPolicy: IfNotPresent
-          env:
-            - name: ANSIBLE_GATHERING
-              value: explicit
-          lifecycle:
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:34]: do not remove lifecycle

File: testpmd-operator/config/manager/manager.yaml [hunk 6]
```diff
@@ -1,63 +1,104 @@
...
-            periodSeconds: 10
-          # operator-sdk is not implementing startup probe up to now, using liveness probe endpoint
-          # while this feature is developed
-          startupProbe:
-            httpGet:
-              path: /healthz
-              port: 6789
-            initialDelaySeconds: 5
-            periodSeconds: 10
-          terminationMessagePolicy: FallbackToLogsOnError
```
Comment (ramperher) [inline testpmd-operator/config/manager/manager.yaml:61]: do not remove terminationMessagePolicy

File: testpmd-operator/config/rbac/kustomization.yaml [hunk 7]
```diff
@@ -12,7 +12,7 @@ resources:
 # Comment the following 4 lines if you want to disable
 # the auth proxy (https://github.com/brancz/kube-rbac-proxy)
 # which protects your /metrics endpoint.
-#- auth_proxy_service.yaml
-#- auth_proxy_role.yaml
-#- auth_proxy_role_binding.yaml
-#- auth_proxy_client_clusterrole.yaml
+- auth_proxy_service.yaml
```
Comment (ramperher) [inline testpmd-operator/config/rbac/kustomization.yaml:15]: I'd keep these four resources commented

File: testpmd-operator/config/rbac/role.yaml [hunk 8]
```diff
@@ -14,7 +14,6 @@ rules:
       - pods
       - pods/exec
       - pods/log
-      - configmaps
```
Comment (ramperher) [inline testpmd-operator/config/rbac/role.yaml:17]: do not delete this, it may be probably needed

File: testpmd-operator/config/rbac/role.yaml [hunk 9]
```diff
@@ -54,49 +54,4 @@ rules:
       - patch
       - update
       - watch
-  - apiGroups:
```
Comment (ramperher) [inline testpmd-operator/config/rbac/role.yaml:57]: do not delete all this, they are required

File: testpmd-operator/config/rbac/role.yaml [hunk 10]
```diff
@@ -54,49 +54,4 @@ rules:
...
-      - rolebindings
-    verbs:
-      - create
-      - delete
-      - get
-      - list
-      - patch
-      - update
-      - watch
-# +kubebuilder:scaffold:rules
```
Comment (ramperher) [inline testpmd-operator/config/rbac/role.yaml:102]: this line is correctly removed, do not re-include it

File: testpmd-operator/config/samples/examplecnf_v1_testpmd.yaml [hunk 11]
```diff
@@ -1,11 +1,12 @@
...
 metadata:
+  labels:
+    app.kubernetes.io/name: testpmd
+    app.kubernetes.io/instance: testpmd-sample
+    app.kubernetes.io/part-of: testpmd-operator-new
+    app.kubernetes.io/managed-by: kustomize
+    app.kubernetes.io/created-by: testpmd-operator-new
   name: testpmd-sample
 spec:
-  image_testpmd: registry.redhat.io/openshift4/dpdk-base-rhel8:v4.6.3-8
```
Comment (ramperher) [inline testpmd-operator/config/samples/examplecnf_v1_testpmd.yaml:6]: you can keep this spec, it's just a sample

File: testpmd-operator/roles/testpmd/defaults/main.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+---
+# defaults file for TestPMD
```
Comment (ramperher) [inline testpmd-operator/roles/testpmd/defaults/main.yml:2]: include this line in main.yaml.in at the beginning of the file

File: testpmd-operator/roles/testpmd/templates/.placeholder
Comment (ramperher) [inline testpmd-operator/roles/testpmd/templates/.placeholder:1]: this can be deleted

**Review by ramperher — CHANGES_REQUESTED** [review]
Some comments, comparing with https://github.com/openshift-kni/example-cnf/pull/53:

- Include in the README the `operator-sdk` commands executed to create the project structure
- roles/testpmd/vars/main.yml file is missed (it's just an empty file but just to have the same than in testpmd-lb-operator)

File: testpmd-operator/Makefile [hunk 13]
```diff
@@ -86,57 +134,64 @@ operator-push:
 	${CONTAINER_CLI} push ${REL_IMG}; \
 	fi
 
-PATH  := $(PATH):$(PWD)/bin
 OS    = $(shell uname -s | tr '[:upper:]' '[:lower:]')
 ARCH  = $(shell uname -m | sed 's/x86_64/amd64/')
```
[the remark this answers is not in the document]
Comment (capolrik) [reply testpmd-operator/Makefile:138]: Are we really supporting arm based system? I can add the line to make it more standard but we don't want to mislead people with false hope.

[the remark this answers is not in the document]
Comment (ramperher) [reply testpmd-operator/Makefile:138]: it's just to have the same it's generated from operator-sdk v1.33.0, I'm not sure we support ARM right now, we may probably check in the future if it were used at some point in a test

File: testpmd-operator/config/manager/manager.yaml [hunk 14]
```diff
@@ -1,63 +1,104 @@
...
       containers:
-        - name: manager
-          securityContext:
-            allowPrivilegeEscalation: false
-          args:
-            - "--enable-leader-election"
-            - "--leader-election-id=testpmd-operator"
-            - "--zap-encoder=console"
-          image: controller:latest
-          imagePullPolicy: IfNotPresent
```
[the remark this answers is not in the document]
Comment (capolrik) [reply testpmd-operator/config/manager/manager.yaml:30]: If the image tag is :latest, the imagePullPolicy will be automatically set to Always. If the image tag isn't :latest, the imagePullPolicy will be automatically set to IfNotPresent. So it should not change the behavior. I'll add it to remain consistent with the rest of the code.

File: testpmd-operator/config/rbac/role.yaml [hunk 15]
```diff
@@ -99,4 +100,3 @@ rules:
       - patch
       - update
       - watch
```
[the remark this answers is not in the document]
Comment (capolrik) [reply testpmd-operator/config/rbac/role.yaml:102]: Sorry, I thought it was here as reference. I have added it again.

**Comment by ramperher** [comment]
Change looks good, I'll accept when having a green job with a full example-cnf deployment and testing

**Comment by ramperher** [comment]
it's weird but it's complaining about this script: https://github.com/dci-labs/example-cnf-config/blob/master/testpmd/hooks/pre-run.yml#L29-L33 , which mainly retrieves the operators that are modified to retrieve the correct SHA. It's like it's not able to see the change in testpmd-operator?

**Comment by capolrik** [comment]
I fix the dependency issue: the pip install cmd was not run under the correct USER in the dockerfile.

**Review by ramperher — APPROVED** [review]
LGTM, thanks Charles for all the great work here!

Let's wait for PR54 to build the example-cnf component before merging this one, to avoid conflicts: https://github.com/openshift-kni/example-cnf/actions/runs/8001713950

**Pull request merged**